File: dyn_mix_cop/fit_markov_dyn_mix_copula.r

source("markov_dyn_mix_copula.r")

library(copula)
# library(egpd)   # load if installed

# ============================================================
# 1. LATENT MIXTURE MODEL (same as your sampler)
# ============================================================

# Gaussian tail copula density
d_tail_cop <- function(u_star, v_star, rho) {
  cop <- normalCopula(param = rho, dim = 2)
  dCopula(cbind(u_star, v_star), copula = cop)
}

# Gumbel bulk copula density
d_bulk_cop <- function(u_star, v_star, alpha) {
  cop <- gumbelCopula(param = alpha, dim = 2)
  dCopula(cbind(u_star, v_star), copula = cop)
}

# Weight function
pi_weight <- function(u_star, v_star, theta) {
  (u_star * v_star)^theta
}

# Unnormalized latent mixture density
c_star_unnorm <- function(u_star, v_star, theta, rho, alpha) {
  w <- pi_weight(u_star, v_star, theta)
  w * d_tail_cop(u_star, v_star, rho) +
    (1 - w) * d_bulk_cop(u_star, v_star, alpha)
}

# ============================================================
# 2. BUILD THE INDUCED COPULA OBJECT FOR A GIVEN psi
#    psi = (theta, rho, alpha)
# ============================================================

build_markov_copula <- function(theta, rho, alpha,
                                grid_size = 120,
                                eps = 1e-6) {
  # ----------------------------------------------------------
  # 2.1 Normalization constant K
  # ----------------------------------------------------------
  K_val <- integrate(
    function(v) {
      sapply(v, function(vv) {
        integrate(
          function(u) c_star_unnorm(u, vv, theta, rho, alpha),
          lower = eps, upper = 1 - eps,
          rel.tol = 1e-5
        )$value
      })
    },
    lower = eps, upper = 1 - eps,
    rel.tol = 1e-5
  )$value

  # Normalized latent joint density c*(u*,v*)
  c_star <- function(u_star, v_star) {
    c_star_unnorm(u_star, v_star, theta, rho, alpha) / K_val
  }

  # ----------------------------------------------------------
  # 2.2 Latent marginal density f_{U*}(u*)
  # ----------------------------------------------------------
  f_u_star <- function(s) {
    sapply(s, function(ss) {
      integrate(
        function(v) c_star(ss, v),
        lower = eps, upper = 1 - eps,
        rel.tol = 1e-5
      )$value
    })
  }

  # ----------------------------------------------------------
  # 2.3 Build monotone spline for F_{U*} and inverse
  # ----------------------------------------------------------
  u_star_grid <- seq(eps, 1 - eps, length.out = grid_size)

  f_grid <- f_u_star(u_star_grid)

  # Numerically integrate marginal density over the grid
  # cumulative trapezoid
  du <- diff(u_star_grid)
  mid_area <- 0.5 * (f_grid[-1] + f_grid[-length(f_grid)]) * du
  F_grid <- c(0, cumsum(mid_area))

  # rescale to [0,1] to absorb numerical integration drift
  F_grid <- F_grid / max(F_grid)

  # strictly increasing safeguard
  F_grid <- pmin(pmax(F_grid, 0), 1)
  for (i in 2:length(F_grid)) {
    if (F_grid[i] <= F_grid[i - 1]) {
      F_grid[i] <- min(1, F_grid[i - 1] + 1e-10)
    }
  }

  F_u_star_spline <- splinefun(u_star_grid, F_grid, method = "monoH.FC")
  F_u_star_inv_spline <- splinefun(F_grid, u_star_grid, method = "monoH.FC")

  # Safe wrappers
  F_u_star <- function(u_star) {
    out <- F_u_star_spline(pmin(pmax(u_star, eps), 1 - eps))
    pmin(pmax(out, eps), 1 - eps)
  }

  F_u_star_inv <- function(u) {
    out <- F_u_star_inv_spline(pmin(pmax(u, eps), 1 - eps))
    pmin(pmax(out, eps), 1 - eps)
  }

  # ----------------------------------------------------------
  # 2.4 Copula density on the observed uniform scale
  #
  # c(u,v) = c*(u*,v*) / [f_{U*}(u*) f_{U*}(v*)]
  # ----------------------------------------------------------
  d_copula <- function(u, v, log = FALSE) {
    u <- pmin(pmax(u, eps), 1 - eps)
    v <- pmin(pmax(v, eps), 1 - eps)

    u_star <- F_u_star_inv(u)
    v_star <- F_u_star_inv(v)

    fu <- f_u_star(u_star)
    fv <- f_u_star(v_star)

    dens <- c_star(u_star, v_star) / (fu * fv)

    dens <- pmax(dens, .Machine$double.xmin)
    if (log) log(dens) else dens
  }

  list(
    theta = theta,
    rho = rho,
    alpha = alpha,
    K = K_val,
    c_star = c_star,
    f_u_star = f_u_star,
    F_u_star = F_u_star,
    F_u_star_inv = F_u_star_inv,
    d_copula = d_copula
  )
}

# ============================================================
# 3. NEGATIVE LOG-LIKELIHOOD FOR FULL MARKOV MODEL
# ============================================================

# Parameter vector layout:
# par = c(theta, z_rho, z_alpha, margin_par...)
#
# We optimize on unconstrained scales and transform:
#   theta > 0          via exp()
#   rho in (-1,1)      via tanh()
#   alpha > 1          via 1 + exp()
#
# You can adapt bounds if needed.

negloglik_markov_egpd <- function(par, x,
                                  margin_npar,
                                  grid_size = 120,
                                  eps = 1e-6,
                                  penalty = 1e12) {
  # ----------------------------------------------------------
  # 3.1 unpack dependence parameters
  # ----------------------------------------------------------
  theta <- exp(par[1]) # > 0
  rho <- tanh(par[2]) # in (-1,1)
  alpha <- 1 + exp(par[3]) # > 1 for Gumbel

  margin_par <- par[4:(3 + margin_npar)]

  # ----------------------------------------------------------
  # 3.2 evaluate marginal CDF / density
  # ----------------------------------------------------------
  # If EGPD parameters are invalid, the wrappers should error or
  # return non-finite values; we catch that below.
  out <- try(
    {
      u <- egpd::pegpd(x,
        kappa = margin_par[1], sigma = margin_par[2], xi = margin_par[3]
      )
      logf <- egpd::degpd_density(x,
        kappa = margin_par[1], sigma = margin_par[2],
        xi = margin_par[3], log = TRUE
      )

      # protect against exact 0/1
      u <- pmin(pmax(u, eps), 1 - eps)

      # build copula object for current dependence parameters
      cop_obj <- build_markov_copula(theta, rho, alpha,
        grid_size = grid_size,
        eps = eps
      )

      # transition log copula density
      logc <- cop_obj$d_copula(u[-length(u)], u[-1], log = TRUE)

      # full Markov log-likelihood:
      # log f(x1) + sum_{t=1}^{n-1}[ log c(u_t,u_{t+1}) + log f(x_{t+1}) ]
      ll <- sum(logf) + sum(logc)

      if (!is.finite(ll)) stop("Non-finite log-likelihood")
      ll
    },
    silent = TRUE
  )

  if (inherits(out, "try-error")) {
    return(penalty)
  }
  return(-out)
}

# ============================================================
# 4. FITTING ROUTINE
# ============================================================

fit_markov_egpd <- function(x,
                            start_dep,
                            start_margin,
                            grid_size = 120,
                            eps = 1e-6,
                            method = "L-BFGS-B",
                            hessian = TRUE,
                            control = list(maxit = 300, trace = 3)) {
  # start_dep should be on natural scale:
  #   list(theta = ..., rho = ..., alpha = ...)
  #
  # start_margin: vector of EGPD starting values on natural scale

  # unconstrained parametrization for optimizer
  par0 <- c(
    log(start_dep$theta),
    atanh(pmin(pmax(start_dep$rho, -0.999), 0.999)),
    log(start_dep$alpha - 1),
    start_margin
  )

  n_margin <- length(start_margin)

  opt <- optim(
    par = par0,
    fn = negloglik_markov_egpd,
    x = x,
    margin_npar = n_margin,
    grid_size = grid_size,
    eps = eps,
    method = method,
    hessian = hessian,
    control = control
  )

  # back-transform estimates
  est_theta <- exp(opt$par[1])
  est_rho <- tanh(opt$par[2])
  est_alpha <- 1 + exp(opt$par[3])
  est_margin <- opt$par[4:(3 + n_margin)]

  # optional covariance on transformed scale:
  vcov_unconstrained <- NULL
  se_unconstrained <- NULL
  if (!is.null(opt$hessian)) {
    invH <- try(solve(opt$hessian), silent = TRUE)
    if (!inherits(invH, "try-error")) {
      vcov_unconstrained <- invH
      se_unconstrained <- sqrt(diag(invH))
    }
  }

  list(
    convergence = opt$convergence,
    message = opt$message,
    logLik = -opt$value,
    par_unconstrained = opt$par,
    estimates = list(
      theta = est_theta,
      rho = est_rho,
      alpha = est_alpha,
      margin = est_margin
    ),
    hessian = opt$hessian,
    vcov_unconstrained = vcov_unconstrained,
    se_unconstrained = se_unconstrained,
    optim = opt
  )
}

fit <- fit_markov_egpd(
  x = x_chain[1:20],
  start_dep = list(theta = 1.2, rho = 0.3, alpha = 1.4),
  start_margin = c(1, 1, 0.4),   # example EGPD starts
  grid_size = 30,
  method = "BFGS",
  control = list(maxit = 10, trace = 6)
)

fit$estimates
fit$logLik

File: dyn_mix_cop/fit_markov_dyn_mix_copula.jl
# markov_fitting.jl
using QuadGK
using Optim
using ForwardDiff
using Interpolations        # replaces Interpolations.jl
using Distributions
using LinearAlgebra
using CSV, DataFrames

const EPS_DEFAULT = 1e-6

# ================================================================
# 1. COPULA COMPONENTS
# ================================================================

function d_gaussian_copula(u::Float64, v::Float64, rho::Float64)
    rho = clamp(rho, -0.999, 0.999)

    u = clamp(u, 1e-12, 1 - 1e-12)
    v = clamp(v, 1e-12, 1 - 1e-12)

    x = quantile(Normal(), u)
    y = quantile(Normal(), v)

    det = 1 - rho^2
    quad = (x^2 - 2*rho*x*y + y^2) / det

    # numerically stable exponent form
    return exp(-0.5 * quad + 0.5*(x^2 + y^2)) / sqrt(det)
end

function d_gumbel_copula(u::Float64, v::Float64, alpha::Float64)
    α = max(alpha, 1.0 + 1e-6)

    u = clamp(u, 1e-12, 1 - 1e-12)
    v = clamp(v, 1e-12, 1 - 1e-12)

    lu = max(-log(u), 1e-12)
    lv = max(-log(v), 1e-12)

    a = lu^α
    b = lv^α
    s = (a + b)^(1/α)

    # protected logs
    log_lu = log(lu)
    log_lv = log(lv)
    log_sum = log(a + b)

    log_c =
        (α - 1) * (log_lu + log_lv) +
        log(s + α - 1) -
        (1 + 1/α) * log_sum -
        log(u) - log(v) -
        s

    return exp(clamp(log_c, -700, 700))
end

pi_weight(u, v, theta) = begin
    u = clamp(u, 1e-12, 1 - 1e-12)
    v = clamp(v, 1e-12, 1 - 1e-12)
    return clamp((u * v)^theta, 0.0, 1.0)
end

function c_star_unnorm(u::Float64, v::Float64,
                       theta::Float64, rho::Float64, alpha::Float64)

    w  = pi_weight(u, v, theta)

    dt = d_gaussian_copula(u, v, rho)
    db = d_gumbel_copula(u, v, alpha)

    val = w * dt + (1 - w) * db

    return isfinite(val) && val > 0 ? val : 1e-300
end


# ================================================================
# 2. EGPD MARGINS
# ================================================================

function egpd_cdf(x::Float64, kappa::Float64, sigma::Float64, xi::Float64)
    sigma = max(sigma, 1e-12)
    kappa = max(kappa, 1e-12)

    if xi == 0.0
        return (1 - exp(-x/sigma))^kappa
    end

    base = 1 - (1 + xi*x/sigma)^(-1/xi)
    base = max(base, 0.0)

    return base^kappa
end

function egpd_logdens(x::Float64, kappa::Float64, sigma::Float64, xi::Float64)
    if xi == 0.0
        return log(kappa) - log(sigma) - x/sigma + (kappa-1)*log(1 - exp(-x/sigma))
    end
    t = 1 + xi*x/sigma
    t <= 0.0 && return -Inf
    log_gpd_dens = -log(sigma) + (-1/xi - 1)*log(t)
    F = max(1 - exp(-1/xi * log(t)), 1e-300)
    return log(kappa) + log_gpd_dens + (kappa-1)*log(F)
end


# ================================================================
# 3. MARKOV COPULA BUILD  (always Float64 — no AD through here)
# ================================================================

struct MarkovCopula{I1, I2}
    theta  :: Float64
    rho    :: Float64
    alpha  :: Float64
    K      :: Float64
    grid   :: Vector{Float64}
    Fgrid  :: Vector{Float64}
    F_fwd  :: I1    # u*  → u   (FritschCarlson ≈ monoH.FC)
    F_inv  :: I2    # u   → u*
end

function build_markov_copula(theta::Float64, rho::Float64, alpha::Float64;
                              grid_size::Int = 120,
                              eps::Float64   = EPS_DEFAULT,
                              rtol::Float64  = 1e-5)

    K_val, _ = quadgk(eps, 1-eps, rtol=rtol) do v
        quadgk(eps, 1-eps, rtol=rtol) do u
            c_star_unnorm(u, v, theta, rho, alpha)
        end[1]
    end

    u_grid = collect(range(eps, 1-eps, length=grid_size))

    f_grid = [
        quadgk(eps, 1-eps, rtol=rtol) do v
            c_star_unnorm(s, v, theta, rho, alpha)
        end[1] / K_val
        for s in u_grid
    ]

    du     = diff(u_grid)
    F_vals = vcat(0.0, cumsum(0.5 .* (f_grid[1:end-1] .+ f_grid[2:end]) .* du))
    F_vals ./= F_vals[end]
    F_vals  = clamp.(F_vals, 0.0, 1.0)

    for i in 2:length(F_vals)
        F_vals[i] = max(F_vals[i], F_vals[i-1] + 1e-12)
    end

    # FritschCarlsonMonotonicInterpolation = R's monoH.FC
    # extrapolate=true handles boundary queries gracefully
    F_fwd = interpolate(u_grid, F_vals, SteffenMonotonicInterpolation())

    F_inv = interpolate(F_vals, u_grid, SteffenMonotonicInterpolation())

    return MarkovCopula(theta, rho, alpha, K_val, u_grid, F_vals, F_fwd, F_inv)
end


# ================================================================
# 4.  TRANSITION LOG-COPULA DENSITY
# ================================================================

function log_copula_density(mc::MarkovCopula, u::Float64, v::Float64)
    u = clamp(u, EPS_DEFAULT, 1-EPS_DEFAULT)
    v = clamp(v, EPS_DEFAULT, 1-EPS_DEFAULT)

    u_star = clamp(mc.F_inv(u), EPS_DEFAULT, 1-EPS_DEFAULT)
    v_star = clamp(mc.F_inv(v), EPS_DEFAULT, 1-EPS_DEFAULT)

    cstar = c_star_unnorm(u_star, v_star, mc.theta, mc.rho, mc.alpha) / mc.K

    fu = quadgk(EPS_DEFAULT, 1-EPS_DEFAULT) do t
        c_star_unnorm(u_star, t, mc.theta, mc.rho, mc.alpha)
    end[1] / mc.K

    fv = quadgk(EPS_DEFAULT, 1-EPS_DEFAULT) do t
        c_star_unnorm(v_star, t, mc.theta, mc.rho, mc.alpha)
    end[1] / mc.K

    return log(max(cstar, floatmin(Float64))) -
           log(max(fu,    floatmin(Float64))) -
           log(max(fv,    floatmin(Float64)))
end


# ================================================================
# 5.  NEGATIVE LOG-LIKELIHOOD
#     par is always Float64 here — the copula precomputation is
#     intentionally kept outside the AD graph (see Section 6).
# ================================================================

function negloglik(par::Vector{Float64}, x::Vector{Float64};
                   grid_size::Int    = 120,
                   eps::Float64      = EPS_DEFAULT,
                   penalty::Float64  = 1e12)

    theta = exp(par[1])
    rho   = tanh(par[2])
    alpha = 1.0 + exp(par[3])
    kappa, sigma, xi = par[4], par[5], par[6]

    (kappa <= 0.0 || sigma <= 0.0) && return penalty

    logf = egpd_logdens.(x, kappa, sigma, xi)
    any(!isfinite, logf) && return penalty

    u = clamp.(egpd_cdf.(x, kappa, sigma, xi), eps, 1-eps)

    mc = try
        build_markov_copula(theta, rho, alpha;
                             grid_size=grid_size, eps=eps)
    catch e
        @warn "copula failed" e
        return penalty
    end

    logc = [log_copula_density(mc, u[t], u[t+1]) for t in 1:length(u)-1]
    any(!isfinite, logc) && return penalty

    return -(sum(logf) + sum(logc))
end


# ================================================================
# 6.  FITTING
#     The Hessian is computed via finite differences on negloglik
#     (not ForwardDiff) because build_markov_copula contains
#     quadrature with Float64-only internals that block AD.
#     For a 6-parameter model this is 2*6^2 = 72 extra likelihood
#     evaluations — fast and numerically stable.
# ================================================================

function hessian_fd(f, p; h=1e-4)
    n = length(p)
    H = zeros(n, n)
    f0 = f(p)
    for i in 1:n, j in i:n
        ei = zeros(n); ei[i] = h
        ej = zeros(n); ej[j] = h
        if i == j
            H[i,i] = (f(p .+ ei) - 2*f0 + f(p .- ei)) / h^2
        else
            H[i,j] = H[j,i] =
                (f(p .+ ei .+ ej) - f(p .+ ei .- ej) -
                 f(p .- ei .+ ej) + f(p .- ei .- ej)) / (4*h^2)
        end
    end
    return H
end

function fit_markov_egpd(x::Vector{Float64};
                          start_theta::Float64  = 1.5,
                          start_rho::Float64    = 0.3,
                          start_alpha::Float64  = 1.2,
                          start_margin::Vector{Float64} = [2.0, 1.0, 0.2],
                          grid_size::Int        = 120,
                          iterations::Int       = 500,
                          show_trace::Bool      = true)

    par0 = [
        log(start_theta),
        atanh(clamp(start_rho, -0.999, 0.999)),
        log(start_alpha - 1.0),
        start_margin...
    ]

    obj(p) = negloglik(p, x; grid_size=grid_size)

    opt = optimize(obj, par0, LBFGS(),
                   Optim.Options(iterations=iterations,
                                 show_trace=show_trace,
                                 g_tol=1e-6))

    par_hat = Optim.minimizer(opt)

    # finite-difference Hessian (safe with quadrature internals)
    H = hessian_fd(obj, par_hat)

    vcov_mat = nothing
    se_vec   = nothing
    try
        invH = inv(Symmetric(H))
        d    = diag(invH)
        if all(isfinite, d) && all(d .>= 0)
            vcov_mat = invH
            se_vec   = sqrt.(d)
        end
    catch e
        @warn "Hessian inversion failed: $e"
    end

    return (
        par         = par_hat,
        estimates   = (
            theta = exp(par_hat[1]),
            rho   = tanh(par_hat[2]),
            alpha = 1.0 + exp(par_hat[3]),
            kappa = par_hat[4],
            sigma = par_hat[5],
            xi    = par_hat[6],
        ),
        convergence = Optim.converged(opt),
        loglik      = -Optim.minimum(opt),
        hessian     = H,
        vcov        = vcov_mat,
        se          = se_vec,
    )
end